File: events/2015/11/solution.ts
import { AOCDayResults } from "@repo/utils-ts/returnType";
import { readInputAsArray } from "@repo/utils-ts/readInput";
import { measurePerformance } from "@repo/utils-ts/measurePerformance";

function incrementPassword(password: string): string {
  let alphabet = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"]
  let newPassword = password.substring(0, password.length-1)

  let lastCharacter = password[password.length-1]
  let lastCharacterIndex = alphabet.indexOf(lastCharacter)

  // Special case for single character password
  if (password.length == 1) {
    if (lastCharacterIndex == 25) {
      newPassword = "a"
    } else {
      newPassword = alphabet[lastCharacterIndex+1]
    }
    return newPassword
  }

  if (lastCharacterIndex == 25) {
    // Rollover
    newPassword = incrementPassword(newPassword) + "a"
  } else {
    // No rollover
    newPassword = newPassword + alphabet[lastCharacterIndex+1]
  }

  return newPassword
}

function isPasswordValid(password: string): boolean {
  let alphabet = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"]

  if (password.includes("i") || password.includes("o") || password.includes("l")) {
    return false
  }

  let repeatingLetters = 0
  let lastRepeatIndex = 0
  for (var i = 1; i < password.length; i++) {
    if (password[i] == password[i-1]) {
      // Repeating Letter!
      if (lastRepeatIndex == 0) {
        repeatingLetters += 1
        lastRepeatIndex = i
      } else if (lastRepeatIndex == i -1) {
        // Doesn't count as it overlaps
        continue
      } else {
        // It's a valid repeat
        repeatingLetters += 1
        lastRepeatIndex = i
      }
    }
  }
  if (repeatingLetters < 2) {
    return false
  }

  for (var i = 2; i < password.length; i++) {
    let current_letter_index = alphabet.indexOf(password[i])

    if (current_letter_index < 2) {
      // Can't contain the sequence we need
      continue
    }

    if (password[i-1] == alphabet[current_letter_index-1] && password[i-2] == alphabet[current_letter_index-2]) {
      return true
    }
  }
  return false
}
// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
function part1(input: string[]): string | undefined {
  let currentPassword = input[0];
  let nextPassword = currentPassword
  let isValid = false

  while (!isValid) {
    nextPassword = incrementPassword(nextPassword)
    isValid = isPasswordValid(nextPassword)
  }
  return nextPassword;
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
function part2(input: string[]): string | undefined {
  let currentPassword = part1(input);

  if (currentPassword == undefined) {
    return undefined
  }

  let nextPassword = currentPassword
  let isValid = false

  while (!isValid) {
    nextPassword = incrementPassword(nextPassword)
    isValid = isPasswordValid(nextPassword)
  }
  return nextPassword;
}


// ----------------------------------------------------------------------------------------------------
// | Solve the puzzle
// ----------------------------------------------------------------------------------------------------
function solve() {
  // Get the arguments
  const args = process.argv.slice(2);
  const fileName = args[0] || "input.txt";

  // Start the timer
  const startTimeStamp = new Date(Date.now()).toISOString();
  const startTime = performance.now();

  // Read the input as an array of strings
  const input = readInputAsArray(fileName);

// Run the parts
const p1Result = measurePerformance(() => part1(input));
const p2Result = measurePerformance(() => part2(input));

// End the timer
const endTimeStamp = new Date(Date.now()).toISOString();
const endTime = performance.now();
const duration = endTime - startTime;

// Return the results
return {
  Year: 2015,
  Day: 11,
  Part1: p1Result,
  Part2: p2Result,
  Duration: Math.round(duration * 1000000), // Convert to nanoseconds
  Timestamp: {
    Start: new Date(startTimeStamp).toISOString(),
    End: new Date(endTimeStamp).toISOString(),
  },
} as AOCDayResults;
  
}
const results = solve();
console.log(JSON.stringify(results, null, 2));

File: events/2015/11/solution.rs
use aoc_utils::read_input_as_array;
use aoc_utils::measure_performance;
use aoc_utils::AOCDayResults;
use aoc_utils::AOCTimestamp;
use chrono::{DateTime, Utc};

fn increment_password(input: String) -> String {
  let alphabet: [&str; 26] = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"];
  let mut new_password = input.clone();
  new_password.pop();

  let last_character = input.chars().last().unwrap();
  let last_character_index = alphabet.iter().position(|&r| r == last_character.to_string()).unwrap();

  if input.chars().count() == 1 {
    if last_character_index == 25 {
      return "a".to_string()
    } else {
      let ret = alphabet[last_character_index+1];
      return ret.to_string();
    }
  }

  if last_character_index == 25 {
    return increment_password(new_password) + "a"
  } else {
    return new_password + alphabet[last_character_index+1]
  }
}

fn is_password_valid(password: String) -> bool {
  let alphabet: [&str; 26] = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"];
  let pass: Vec<char> = password.chars().collect();
  let password_length = password.chars().count();

  if password.contains("i") || password.contains("o") || password.contains("l") {
    return false
  }

  let mut repeating_letters = 0;
  let mut last_repeat_index = 0;
  for i in 1..password_length {
    if pass[i] == pass[i-1] {
      if last_repeat_index == 0 {
        repeating_letters += 1;
        last_repeat_index = i;
      } else if last_repeat_index == i - 1{
        continue;
      } else {
        last_repeat_index = i;
        repeating_letters += 1;
      }
    }
  }
  if repeating_letters < 2 {
    return false
  }

  for i in 2..password_length {
    let current_letter_index = alphabet.iter().position(|&r| r == pass[i].to_string()).unwrap();

    if current_letter_index < 2 {
      continue;
    }

    if pass[i-1].to_string() == alphabet[current_letter_index-1] && pass[i-2].to_string() == alphabet[current_letter_index-2] {
      return true
    }
  }
  return false
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
fn part1(input: &Vec<String>) -> String {
  let current_password = &input[0];
  let mut next_password = current_password.clone();
  let mut is_valid = false;

  while !is_valid {
    next_password = increment_password(next_password);
    is_valid = is_password_valid(next_password.clone());
  }

  next_password
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
fn part2(input: &Vec<String>) -> String {
  let current_password = part1(input);
  let mut next_password = current_password.clone();
  let mut is_valid = false;

  while !is_valid {
    next_password = increment_password(next_password);
    is_valid = is_password_valid(next_password.clone());
  }

  next_password
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
fn main() {
  // Read the arguments
  let args = std::env::args().collect::<Vec<String>>();
  let default_file_name = String::from("input.txt");
  let file_name = args.get(1).unwrap_or(&default_file_name);

  // Start the timer
  let start_timestamp = std::time::SystemTime::now();
  let start_time = std::time::Instant::now();
  
  // Read the input
  let input = read_input_as_array(file_name).unwrap();

  // Run the Parts
  let p1_result = measure_performance(part1, &input);
  let p2_result = measure_performance(part2, &input);

  // Stop the timer
  let end_time = std::time::Instant::now();
  let end_timestamp = std::time::SystemTime::now();
  let duration = end_time.duration_since(start_time).as_nanos();

  let results = AOCDayResults {
    year: 2015,
    day: 11,
    part1: p1_result,
    part2: p2_result,
    duration: duration,
    timestamp: AOCTimestamp {
      start: DateTime::<Utc>::from(start_timestamp).to_rfc3339(),
      end: DateTime::<Utc>::from(end_timestamp).to_rfc3339(),
    },
  };

  let json_results = serde_json::to_string(&results).unwrap();
  println!("{}", json_results);
}